src/core/mac/sub_mac.cpp

```diff
@@ -357,9 +357,10 @@ void SubMac::StartCsmaBackoff(void)
         if (ShouldHandleTransmitTargetTime())
         {
             if (Time(static_cast<uint32_t>(otPlatRadioGetNow(&GetInstance()))) <
-                Time(mTransmitFrame.mInfo.mTxInfo.mTxDelayBaseTime) + mTransmitFrame.mInfo.mTxInfo.mTxDelay)
+                Time(mTransmitFrame.mInfo.mTxInfo.mTxDelayBaseTime) + mTransmitFrame.mInfo.mTxInfo.mTxDelay -
+                    kCcaSampleInterval)
```

Comment:
I agree that a margin should be added here to avoid too late requesting of precise 802.15.4 transmissions. However, I think the duration of this delays should depend on two factors:

1. If CCA is expected before the frame being transmitted (to be on a safe side, we can assume that CCA is always expected, like here).
2. What radio is being used. For sure this margin should be a little longer for nRF5340 than for nRF52840 (because of serialization of this call between CPUs). I expect other vendors would require different timings, especially when the turnaround time differs.

How about adding a function to radio platform that would report how long delay is required?

Reply:
For platform which supports delay transmission(`OT_RADIO_CAPS_TRANSMIT_TIMING`), `ShouldHandleTransmitTargetTime()` will return false and the frame will be delivered to Radio driver directly with `mTxDelayBaseTime` and `mTxDelay` fields set. The frame will arrives at Radio driver about `OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_US`(2000ms) ahead of the target transmission time. Radio driver can adjust the delay according to CCA, Tx ramp up, Tx setup time and software delay by itself. Just like what is done in Nordic platform `nrf_802154_delayed_trx_transmit()`.

The handling here is for the platform which does not support delay transmission(`OT_RADIO_CAPS_TRANSMIT_TIMING`). It will rely on `SubMac` to do the job. I agree for this case, we should define some platform specific delay to compensate the time lag between `otPlatRadioTransmit` and when the first symbol of the frame is finally sent.

Reply:
> For platform which supports delay transmission(`OT_RADIO_CAPS_TRANSMIT_TIMING`), `ShouldHandleTransmitTargetTime()` will return false

I wasn't aware of that. I thought that this code is for platforms supporting delayed transmissions. Thank you for explanation.

Reply:
Thanks @hubertmis for the review. I will mark this comment and the next one resolved for this PR. For enhancements to support the platforms which does not have delayed transmission, I will do it as a separate PR.

For the timing differences between nRF5340 and nRF52840, if they both supports `OT_RADIO_CAPS_TRANSMIT_TIMING`, then I think you can define different `TX_SETUP_TIME` or `TX_RAMP_UP_TIME` in your drivers for different chips.

Reply:
> For the timing differences between nRF5340 and nRF52840, if they both supports `OT_RADIO_CAPS_TRANSMIT_TIMING`, then I think you can define different `TX_SETUP_TIME` or `TX_RAMP_UP_TIME` in your drivers for different chips.

On the 802.15.4 radio driver side, we'll of course trim constants to make sure that everything works correctly. However, I think that there might be required some changes in OT, like selecting `OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_US` value depending on a SoC. nRF5340 may require a little longer time than nRF52.

But we can handle it later, when we introduce nRF53 support.

Reply:
Yes, if default 2ms is still not enough for nRF53, then this value should be set larger.

BTW, we found one note [here](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.thread_zigbee.v2.0.0%2Fgroup__nrf__802154__delayed__trx.html) which is different from the function description in nrf_802154.h. Is this note still valid?

> Delayed transmission does not timeout waiting for ACK automatically. Waiting for ACK shall be timed out by the next higher layer or the ACK timeout module. The ACK timeout timer shall start when the nrf_802154_tx_started function is called.

As Nordic driver claims the capability of `OT_RADIO_CAPS_ACK_TIMEOUT`, so OT stack does not start a timer to time out a transmission and will wait for callback from Radio for a transmission.

Reply:
The limitation you found is not present anymore. The radio driver times out waiting for ACKs by itself for both delayed and typical TX operations.

Reply:
Thanks for the confirmation.